modules/calib3d/src/solvepnp.cpp

```diff
@@ -201,6 +203,11 @@ bool solvePnPRansac(InputArray _opoints, InputArray _ipoints,
 {
     CV_INSTRUMENT_REGION();
 
+    if (flags == USAC_DEFAULT || flags == USAC_PARALLEL)
+        return usac::solvePnPRansac(_opoints, _ipoints, _cameraMatrix, _distCoeffs,
```

Comment:
Wondering if it is planned to improve the whole `solvePnPRansac` logic or if you know a good logic that could be used?

I am referring to this topic: #8782

---

From memory and from reading quickly what I wrote:

- default behavior uses EPnP to compute the solution for the Minimal Sample Sets (MSS) using 5 points
- I thought that the point of RANSAC is to use the minimal number of points so 4 points for pose estimation
- but probably it was done like this because EPnP works well with good number of points, wondering if it is starting from 6 points EPnP works well?
- good luck to understand the behavior of the function if you change the PnP method

Also:

- intuitively, I think P3P methods should be used for MSS step since they can produce at most 4 possible solutions with 3 points
- so what about using AP3P for MSS? Gao P3P will give bad results due to not handling all [degenerate cases](https://github.com/opencv/opencv/blob/dbab8d8c3897040f134af7195e130dbfcdc2c065/modules/calib3d/src/p3p.cpp#L221)
- in general, I would add explicitly the PnP method for MSS and for the final pose computation
- because in the present case it gives something like this:

![image](https://user-images.githubusercontent.com/8229425/89507204-7d143d00-d7cc-11ea-938b-bab8beb32a4a.png)

Reply:
This is what it should look like in practice: [AP3P vs P3P](https://www.youtube.com/watch?v=X5h_4okDzyI)

Reply:
I haven't planned to change a whole logic of solvePnP. My goal is about improving RANSAC which will solve PnP too. I added classical P3P solver which seems a bit different from the one in OpenCV. I am not very familiar with whole family of PnP solvers but I think in case of P3P even if solution is not precise enough (e.g., comparing to AP3P) then it will be polished in local optimization step of RANSAC by some DLS or DLT solver. My new idea that I proposed to Vadim is adding P6P solver which does not assume knowledge of intrinsic matrix which is often the case.

Reply:
Indeed, there should be no need to have a super precise P3P for the the Minimal Sample Sets step since the final pose will be refined using all the data in the best consensus set (assuming USAC works similarly to the classical RANSAC).

Just curious to know if you are using P3P with 4 points to have a unique solution or if you are using 3 points (that can produce up to 4 solutions, probably some can be eliminated using positive t_z)?

---

> I haven't planned to change a whole logic of solvePnP.

Yes, my comment was actually more general and I realise it is not really related to this GSoC work.

Probably the answer I would have is that now that there is USAC method that should work better than RANSAC, there is no need to fix the logic inside `solvePnpRansac`. I personnally disagree with this.

Reply:
I use the one which produces up to 4 solutions. However, solving 4-degrees polynomial can give you 0, 2 or 4 real roots. So, I experimented a bit and found that mostly it has only 2 real roots and average number of models after applying different verifications is around 1.38